src/lib/array.jl

```diff
@@ -957,3 +959,15 @@ end
   back(Δ::AbstractArray) = (nothing, getindex.(_back.(Δ), 1))
   return Fill(y, size(r)), back
 end
+
+# Sparse Arrays
+
+@adjoint function SparseMatrixCSC{T,N}(arr) where {T,N}
+  SparseMatrixCSC{T,N}(arr), Δ -> (collect(Δ),)
+end
+@adjoint diagm(x) = diagm(x), Δ -> (diag(Δ), )
```

Comment:
Perhaps consider restricting the type here a bit? For example, this adjoint is probably innappropriate for `Fill`s since `diag(Δ)` will presumably be dense.

Reply:
Shouldn't  `diagm` fail in that case?

Reply:
Well, for example,
```julia
julia> diagm(Fill(5.0, 3))
3×3 Array{Float64,2}:
 5.0  0.0  0.0
 0.0  5.0  0.0
 0.0  0.0  5.0
```
seems to be well-defined.

Reply:
Would you recommend `DenseArray`?

Reply:
Actually we need it for the sparse arrays, so I'm not quite sure

Reply:
That would work for me!

Reply:
Ah, yes, good point. Maybe restrict it to sparse arrays for now then, and expand if necessary later?